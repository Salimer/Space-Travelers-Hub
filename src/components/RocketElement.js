import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelReserveRocket } from '../redux/rockets/rocketsSlice';

const RocketElement = ({ rocket }) => {
  const dispatch = useDispatch();
  const handleClick = (e, id) => {
    e.preventDefault();
    if (rocket.reserved === false) {
      dispatch(reserveRocket({ id }));
    } else {
      dispatch(cancelReserveRocket({ id }));
    }
  };
  return (
    <Li>
      <div className="img">
        <img src={rocket.flickr_images[0]} alt="rocket" />
      </div>
      <div className="details">
        <span>{rocket.rocket_name}</span>
        <p>
          {rocket.reserved === true && <span className="reserved">Reserved</span>}
          {rocket.description}
        </p>
        <button onClick={(e) => { handleClick(e, rocket.id); }} className="reserve-btn" type="button">{rocket.reserved === true ? 'Cancel Reservation' : 'Reserve Rocket'}</button>
      </div>
    </Li>
  );
};

export default RocketElement;

RocketElement.propTypes = {
  rocket: PropTypes.shape({
    id: PropTypes.number.isRequired,
    rocket_name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    flickr_images: PropTypes.arrayOf(PropTypes.string).isRequired,
    rocket_id: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
  }).isRequired,
};

const Li = styled.li`
display: flex;
border: 3px solid black;
margin: 1rem;
border-radius: 1rem;
padding: 1rem;

.img {
  display: flex;
  align-items: center;

  img {
    width: 10rem;
    padding-left: 1rem;
  }
}

.details {
  display: flex;
  gap: 1rem;
  flex-direction: column;
  margin-left: 2rem;
  justify-content: center;

  .reserved {
    font-size: 10px;
    border-radius: .2rem;
    margin-right:.5rem;
    padding: 0.2rem;
    background-color: green;
    color: white;
  }

  .reserve-btn {
    max-width: fit-content;
    padding: .5rem;
    background-color: blue;
    cursor: pointer;
    border-radius: .4rem;
    color: white;
    border: transparent;
  }
}
`;
