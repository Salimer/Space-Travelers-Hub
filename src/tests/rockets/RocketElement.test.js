import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import RocketElement from '../../components/RocketElement';
import rocketsReducer from '../../redux/rockets/rocketsSlice';

const store = configureStore({
    reducer: {
      rockets: rocketsReducer, 
    },
  });

test('renders a rocket item correctly', () => {
    const rockets = {
        id: 1,
          rocket_id: 'u1',
          rocket_name: 'Falcon 9',
          description: 'A two-stage rocket designed and manufactured by SpaceX.',
          flickr_images: ['falcon9.jpg'],
          reserved: false,
        };

    const { container } = render (
        <Provider store={store}>
            <RocketElement rocket={rockets}/>
        </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByText('Falcon 9')).toBeInTheDocument();
    expect(screen.getByText('A two-stage rocket designed and manufactured by SpaceX.')).toBeInTheDocument();
    expect(screen.getByText('Reserve Rocket')).toBeInTheDocument();
})

