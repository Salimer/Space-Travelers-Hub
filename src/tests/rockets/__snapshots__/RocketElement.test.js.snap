// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a rocket item correctly 1`] = `
<li
  class="sc-bgqQcB jHJsdY"
>
  <div
    class="img"
  >
    <img
      alt="rocket"
      src="falcon9.jpg"
    />
  </div>
  <div
    class="details"
  >
    <span>
      Falcon 9
    </span>
    <p>
      A two-stage rocket designed and manufactured by SpaceX.
    </p>
    <button
      class="reserve-btn"
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</li>
`;
